cirq-core/cirq/transformers/merge_single_qubit_gates.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Transformer passes to combine adjacent single-qubit rotations."""
+
+from typing import Optional, Callable, List, TYPE_CHECKING
+
+import numpy as np
+
+from cirq import ops, protocols, circuits
+from cirq.transformers.analytical_decompositions import single_qubit_decompositions
+from cirq.transformers import transformer_api, transformer_primitives
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@transformer_api.transformer
+def merge_single_qubit_gates(
+    circuit: 'cirq.AbstractCircuit',
+    *,
+    context: Optional['cirq.TransformerContext'] = None,
+    rewriter: Optional[Callable[[List[ops.Operation]], ops.OP_TREE]] = None,
+    synthesizer: Optional[Callable[[ops.Qid, np.ndarray], ops.OP_TREE]] = None,
+) -> 'cirq.Circuit':
+    """Merges adjacent single qubit unitaries in the given circuit.
+
+    Uses rewriter or synthesiser to convert a connected component of single qubit unitaries
+    into a more desirable form. At-most one of rewriter or synthesiser should be specified
+    at a time. If both are not specified, connected components are replaced by a single
+    `cirq.MatrixGate` containing unitary of the merged component.
+
+    Args:
+        circuit: Input circuit to transform. It will not be modified.
+        context: `cirq.TransformerContext` storing common configurable options for transformers.
+        rewriter: Specifies how to merge runs of single-qubit operations into a more desirable
+            form. Takes a list of operations and produces a list of operations.
+        synthesizer: A special kind of rewriter that operates purely on the unitary matrix of the
+            intended operation. Takes a qubit and a unitary matrix and returns a list of operations.
+
+    Returns:
+        Copy of the transformed input circuit.
+
+    Raises:
+        ValueError: If both rewriter and synthesizer were specified.
+    """
+    if rewriter is not None and synthesizer is not None:
+        raise ValueError("Can't specify both rewriter and synthesizer.")
+    merged_circuit_op_tag = "_merged_single_qubit_gates_component"
```

Comment:
No, the tag should be local because it's used as an implementation detail to identify circuit operations which were created by the internal merge and given them special treatment while passing arguments to `rewriter`. Note that this tag is never applied to any operation of the final circuit being returned from the function.